### ISSUE TITLE
caf::detail::beacon is not dereferenced properly when actor_system shuts down

### ISSUE THREAD

**Issue by shariarriday** [actor-framework/actor-framework#1954]
The following example tries to create a scenario where the `actor_system` is shut down when the execution is blocked by `res->first.get()`. This example is not completely deterministic but reproduces the problem often. 

Also, "http://localhost:9880" should be able to handle requests and send responses.

```
#include "caf/net/http/client.hpp"
#include "caf/net/http/with.hpp"
#include "caf/net/middleman.hpp"

#include "caf/actor_system.hpp"
#include "caf/actor_system_config.hpp"
#include "caf/caf_main.hpp"

#include <memory>
#include <utility>

using namespace std::literals;

namespace http = caf::net::http;
namespace ssl = caf::net::ssl;
using namespace caf;

int caf_main(caf::actor_system& sys) {
  std::thread worker([&sys]() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    sys.~actor_system();
  });
  while (true) {
    namespace http = caf::net::http;
    namespace ssl = caf::net::ssl;
    caf::uri uri;
    if (auto err = parse("http://localhost:9880", uri)) {
      sys.println("*** failed to parse URI: {} ", err);
      return EXIT_FAILURE;
    }
    auto res = http::with(sys)
                 .context(ssl::context::enable(false).and_then(
                   ssl::emplace_client(ssl::tls::v1_2)))
                 .connect(uri)
                 .connection_timeout(500ms)
                 .max_retry_count(0)
                 .add_header_fields(std::map<std::string, std::string>{
                   {"content-type", "application/json"}})
                 .request(http::method::post, "{\"test\": \"Message\"}");
    if (res && res->first.get()) {
      continue;
    } else {
      printf("Failed to send request");
      return EXIT_SUCCESS;
    }
  }
}

CAF_MAIN(caf::net::middleman)

```

**Comment by riemass** [comment]
I've had to modify the dtor of `actor_system` so we don't run it twice simultaneously (this shouldn't be an issue in real scenarios), and I've added `worker.join()` before returning from `caf_main`. 
  
 ```diff
 actor_system::~actor_system() {
-  impl_->~impl();
-  free(impl_);
+  if (impl_) {
+    impl_->~impl();
+    free(impl_);
+    impl_ = nullptr;
+  }
 }
```

This fixed all the terminations/segfaults of the example I've had. What remains now is that the final shutdown is blocked by the `future.get()` call in the main thread. The actor_system has finished it's dtor, but the app is still handing around on the `cv` in `res->first.get()`. Is this the main issue here you reported or I'm missing something? Just asking for a bit more context so I know in which direction to go.

**Comment by shariarriday** [comment]
> The actor_system has finished it's dtor, but the app is still handing around on the cv in res->first.get()

You are correct, this is the main issue that I reported. I tried modifying the dtors but the `cv` in `res->first.get()` was the main issue. The application was stuck due to it.

**Comment by riemass** [comment]
This is definitely a weird case. The `res->first.get()` gets unblocked with an error value when the destructor is run.  However, the while loop successfully establishes a new connection and creates a new `res`. It's the new `res->first.get()` that was created after the destructor that's blocking the system shutdown. Not sure how this is possible, but the `actor_system` destructor schedules a shutdown event in the `mpx` thread, so it's an asynchronous operation. Those are my findings, I'll have to look into the `mpx` scheduling to find an appropriate workaround.

**Comment by shariarriday** [comment]
@riemass It looks very similar to my finding. The `res->first.get()` that is created after the destructor was the causing the issue and the bug is not 100% reproducible due to this. I have attempted to find an appropriate workaround but could not find a solution. I am not well versed on the `mpx` scheduling, so I am hoping that you will have a better time in finding a solution.
